url2b521cbda7_eobermuhlner_jhuge_tgz-pJ8-ch_obermuhlner_jhuge_example_MeasureHeapJ8/src/ch/obermuhlner/jhuge/converter/CompactConverter.java

```diff
@@ -0,0 +1,657 @@
+			byteBuffer.putInt(value.length);
+			for (int i = 0; i < value.length; i++) {
+				byteBuffer.putDouble(value[i]);
+			}
+			return compactData;
+		}
+		if (clazz == boolean[].class) {
+			boolean[] value = (boolean[]) source;
+			int compactLength = (value.length + 7) / 8;
+			byte[] compactData = new byte[1 + 4 + compactLength];
+			ByteBuffer byteBuffer = ByteBuffer.wrap(compactData);
+			byteBuffer.put(ARRAY_BOOLEAN);
+			byteBuffer.putInt(value.length);
+			byte currentByte = 0;
+			for (int i = 0; i < value.length; i++) {
+				int moduloIndex = i % 8;
+				if (value[i]) {
+					currentByte |= 1 << moduloIndex; 
+				}
+				if (i == 7 || i == value.length - 1) {
+					byteBuffer.put(currentByte);
+					currentByte = 0;
+				}
+			}
+			return compactData;
+		}
+		if (clazz == char[].class) {
+			char[] value = (char[]) source;
+			byte[] compactData = new byte[1 + 4 + value.length * 2];
+			ByteBuffer byteBuffer = ByteBuffer.wrap(compactData);
+			byteBuffer.put(ARRAY_CHAR);
+			byteBuffer.putInt(value.length);
+			for (int i = 0; i < value.length; i++) {
+				byteBuffer.putChar(value[i]);
+			}
+			return compactData;
+		}
+		if (clazz == BigDecimal[].class) {
+			BigDecimal[] value = (BigDecimal[]) source;
+			ByteArrayOutputStream aByteArrayOutputStream = new ByteArrayOutputStream();
+			DataOutputStream stream = new DataOutputStream(aByteArrayOutputStream);
+			try {
+				stream.writeByte(ARRAY_BIGDECIMAL);
+				stream.writeInt(value.length);
+				for (int i = 0; i < value.length; i++) {
+					BigDecimal element = value[i];
+					if (element == null) {
+						stream.writeByte(NULL);
+					} else if (element.getClass() == BigDecimal.class){
+						stream.writeByte(BIGDECIMAL);
+						stream.writeUTF(element.toString());
+					} else {
+						stream.writeByte(SERIALIZED);
+						serialize(element, stream);
+					}
+				}
+				stream.close();
+				return aByteArrayOutputStream.toByteArray();
+			}
+			catch (IOException exception) {
+				// ignore
+			}
+		}
+		if (clazz == String[].class) {
+			String[] value = (String[]) source;
+			ByteArrayOutputStream aByteArrayOutputStream = new ByteArrayOutputStream();
+			DataOutputStream stream = new DataOutputStream(aByteArrayOutputStream);
+			try {
+				stream.writeByte(ARRAY_STRING);
+				stream.writeInt(value.length);
+				for (int i = 0; i < value.length; i++) {
+					String element = value[i];
+					if (element == null) {
+						stream.writeByte(NULL);
+					} else {
+						stream.writeByte(STRING);
+						stream.writeUTF(element);
+					}
+				}
+				stream.close();
+				return aByteArrayOutputStream.toByteArray();
+			}
+			catch (IOException exception) {
+				// ignore
+			}
+		}
+		
+		byte[] serializedData = super.serialize(source);
+		byte[] compactData = new byte[serializedData.length + 1];
+		compactData[0] = SERIALIZED;
+		System.arraycopy(serializedData, 0, compactData, 1, serializedData.length);
+		return compactData;
+	}
+	
+	@Override
+	public T deserialize(byte[] data) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 176 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.